### PR TITLE
Throw an error when the page count or perPage setting is invalid

### LINKED ISSUE #18924: Sharee api pagination makes page 0 and 1 be the same.
When requesting a list of users and groups for sharing and specifying a limit now the page 0 and page 1 are the same.

Having a server with several users and groups.

This request
```
curl -X GET "http://CRE:DENTIALS@HOST/ocs/v1.php/apps/files_sharing/api/v1/sharees?itemType=file&limit=3&page=0" 
```
Returns the same as
```
curl -X GET "http://CRE:DENTIALS@HOST/ocs/v1.php/apps/files_sharing/api/v1/sharees?itemType=file&limit=3&page=1" 
```

Happens the same with negative numbers. (page=-1).

I think it should start in 0 and return a 404 or similar with negative numbers.

@DeepDiver1975 @nickvergessen @rullzer

### PR DESCRIPTION
**Pull request by nickvergessen**
Instead of duplicating pages tell the user that he was wrong

Fix #18924 

@DeepDiver1975 @rullzer @SergioBertolinSG

### CODE REVIEW

**Comment by karlitschek** [comment]
makes sense :+1:

File: apps/files_sharing/api/sharees.php [hunk 1]
```diff
@@ -291,8 +291,15 @@ protected function getRemote($search) {
...
-		$page = !empty($_GET['page']) ? max(1, (int) $_GET['page']) : 1;
-		$perPage = !empty($_GET['perPage']) ? max(1, (int) $_GET['perPage']) : 200;
+		$page = isset($_GET['page']) ? (int) $_GET['page'] : 1;
+		$perPage = isset($_GET['perPage']) ? (int) $_GET['perPage'] : 200;
+
+		if ($perPage <= 0) {
+			return new \OC_OCS_Result(null, 401, 'Invalid perPage argument');
+		}
+		if ($page <= 0) {
+			return new \OC_OCS_Result(null, 402, 'Invalid page');
```
Comment (DeepDiver1975) [inline apps/files_sharing/api/sharees.php]: 402 Payment Required

will that really help? :speak_no_evil:

File: apps/files_sharing/api/sharees.php [hunk 2]
```diff
@@ -291,8 +291,15 @@ protected function getRemote($search) {
 	public function search() {
 		$search = isset($_GET['search']) ? (string) $_GET['search'] : '';
 		$itemType = isset($_GET['itemType']) ? (string) $_GET['itemType'] : null;
-		$page = !empty($_GET['page']) ? max(1, (int) $_GET['page']) : 1;
-		$perPage = !empty($_GET['perPage']) ? max(1, (int) $_GET['perPage']) : 200;
+		$page = isset($_GET['page']) ? (int) $_GET['page'] : 1;
+		$perPage = isset($_GET['perPage']) ? (int) $_GET['perPage'] : 200;
+
+		if ($perPage <= 0) {
+			return new \OC_OCS_Result(null, 401, 'Invalid perPage argument');
```
Comment (DeepDiver1975) [inline apps/files_sharing/api/sharees.php]: 401 Unauthorized 

????

File: apps/files_sharing/api/sharees.php (hunk 1 above)
Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: It's OCS not HTTP?

Missing itemType is code 400...

I can change them all if you tell me the desired value

Comment (DeepDiver1975) [inline apps/files_sharing/api/sharees.php]: OCS error codes have to match up http status codes by now

Comment (nickvergessen) [inline apps/files_sharing/api/sharees.php]: Okay, fixed to `Http::STATUS_BAD_REQUEST`

**Comment by rullzer** [comment]
:+1:

**Comment by SergioBertolinSG** [comment]
I've been testing this and works fine. 
But if you puts a number in page where there shouldn't be any value it returns 100 and empty results instead of invalid page.
Is this behaviour correct?

This is the result when you request a page too far away
```
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>ok</status>
  <statuscode>100</statuscode>
  <message/>
 </meta>
 <data>
  <exact>
   <users/>
   <groups/>
   <remotes/>
  </exact>
  <users/>
  <groups/>
  <remotes/>
 </data>
</ocs>
```

**Comment by nickvergessen** [comment]
Yes, because that can also happen on page 1 and we also don't know what the last "valid" page would be.

**Comment by SergioBertolinSG** [comment]
OK :+1: then.

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**